code/mob/living/critter/ai/sawflyai.dm

```diff
@@ -0,0 +1,212 @@
+		//	playsound(owncritter, pick(owncritter.beeps), 55, 1) //actively do not use the communalbeep proc
+			owncritter.set_dir(get_dir(owncritter, holder.target)) // since they're probably in range, attack regardless
+			owncritter.hand_range_attack(holder.target, dummy_params)
+			SPAWN(5)
+				owncritter.set_dir(get_dir(owncritter, holder.target)) //follow up
+				owncritter.hand_range_attack(holder.target, dummy_params)*/
+
+/datum/aiTask/prioritizer/sawfly
+	name = "base thinking (should never see this)"
+
+/datum/aiTask/prioritizer/sawfly/New()
+	..()
+	// populate the list of tasks
+	transition_tasks += holder.get_instance(/datum/aiTask/timed/wander, list(holder, src))
+	transition_tasks += holder.get_instance(/datum/aiTask/timed/targeted/sawfly_attack, list(holder, src))
+	transition_tasks += holder.get_instance(/datum/aiTask/sequence/goalbased/sawfly_chase_n_stab, list(holder, src))
+
+/datum/aiTask/prioritizer/sawfly/on_tick()
+	if(isdead(holder.owner))
+		holder.enabled = FALSE
+		walk(holder.owner, 0)
+
+/datum/aiTask/prioritizer/sawfly/on_reset()
+	..()
+	walk(holder.owner, 0)
+
+// Custom behaviour starts here
+
+// Attack code - looks for adjacent targets, tries to stab them
+/datum/aiTask/timed/targeted/sawfly_attack
+	var/found_path = null
+	name = "attacking"
+	minimum_task_ticks = 10
+	maximum_task_ticks = 25
+	var/weight = 10
+	target_range = 1
+	var/list/dummy_params = list("icon-x" = 16, "icon-y" = 16)
+
+/datum/aiTask/timed/targeted/sawfly_attack/evaluate()
+	return weight * score_target(get_best_target(get_targets()))
+
+/datum/aiTask/timed/targeted/sawfly_attack/on_tick()
+	var/mob/living/critter/sawfly/owncritter = holder.owner
+	if(prob(5)) owncritter.communalbeep()
+	walk(owncritter, 0)
+	if(!holder.target)
+		holder.target = get_best_target(get_targets())
+	if(holder.target)
+		var/atom/T = holder.target
+		// if target is dead
+		// fetch a new one if we can
+
+		if(isliving(T))
+			var/mob/living/M = T
+			if(M.health < -50)
+				holder.target = get_best_target(get_targets())
+			if(istype(T, /mob/living/critter/sawfly))
+				holder.target = get_best_target(get_targets())
+		if(!holder.target) //we lost target, re-evaluate tasks
+			holder.interrupt()
+			return
+
+		var/dist = get_dist(owncritter, holder.target)
+		if(dist > target_range)
+			//replaced
+			if(!src.found_path)
+				src.found_path = get_path_to(holder.owner, holder.target, 18, 0)
+			if(src.found_path)
+				holder.move_to_with_path(holder.target, src.found_path, 1)
+				if(prob(20)) walk_rand(src,4)
+				owncritter.set_dir(get_dir(owncritter, holder.target)) //attack regardless
+				owncritter.hand_range_attack(holder.target, dummy_params)
+			else
+				//o no
+				frustration++
+			frustration++ //if frustration gets too high, the task is ended and re-evaluated
+		else
+			owncritter.set_dir(get_dir(owncritter, holder.target))
+			owncritter.hand_range_attack(holder.target, dummy_params)
+			SPAWN(10)
```

Comment:
This feels bad, and is likely to result in some janky behaviour if the task gets interrupted. If you aren't clearing holder target then it should just attack the same target again on the next ai tick